### PR TITLE
Create exercise 12

### PR DESCRIPTION
**Pull request by maryx**
- Adds error UI
- I changed the error UI in the master unit converter app too

Files of note:
unit_converter/unit_converter/lib/category.dart
unit_converter/unit_converter/lib/category_tile.dart
unit_converter/unit_converter/lib/category_route.dart

course/12_error/solution_12_error/lib/category.dart
course/12_error/solution_12_error/lib/category_tile.dart
course/12_error/solution_12_error/lib/category_route.dart
course/12_error/solution_12_error/lib/unit_converter.dart

course/12_error/task_12_error/lib/category.dart
course/12_error/task_12_error/lib/category_tile.dart
course/12_error/task_12_error/lib/category_route.dart
course/12_error/task_12_error/lib/unit_converter.dart

This is the last PR for exercises!!

### CODE REVIEW

File: course/12_error/solution_12_error/lib/category.dart [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  /// its color for the UI, units for conversions (e.g. 'Millimeter', 'Meter'),
+  /// and the icon that represents it (e.g. a ruler).
+  const Category({
+    @required this.name,
+    @required this.color,
+    @required this.units,
+    @required this.iconLocation,
+  })  : assert(name != null),
+        assert(color != null),
+        assert(units != null);
```
Comment (yjbanov) [inline course/12_error/solution_12_error/lib/category.dart]: Can `iconLocation` be null?

File: course/12_error/solution_12_error/lib/category_route.dart [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            onTap:
+                _category.name == apiCategory['name'] && _category.units.isEmpty
+                    ? null
+                    : _onCategoryTap,
+          );
+        },
+        itemCount: _categories.length,
+      );
+    } else {
+      // Why do we pass in `_categories.toList()` instead of just `_categories`?
```
Comment (yjbanov) [inline course/12_error/solution_12_error/lib/category_route.dart]: ditto

File: course/12_error/solution_12_error/lib/unit_converter.dart [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+
+  /// Sets the default values for the 'from' and 'to' [Dropdown]s, and the
+  /// updated output value if a user had previously entered an input.
+  void _setDefaults() {
+    setState(() {
+      _fromValue = widget.category.units[0];
+      _toValue = widget.category.units[1];
+    });
+    if (_inputValue != null) {
+      setState(() {
```
Comment (yjbanov) [inline course/12_error/solution_12_error/lib/unit_converter.dart]: ditto

File: course/12_error/task_12_error/lib/category.dart [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  /// its color for the UI, units for conversions (e.g. 'Millimeter', 'Meter'),
+  /// and the icon that represents it (e.g. a ruler).
+  const Category({
+    @required this.name,
+    @required this.color,
+    @required this.units,
+    @required this.iconLocation,
+  })  : assert(name != null),
+        assert(color != null),
+        assert(units != null);
```
Comment (yjbanov) [inline course/12_error/task_12_error/lib/category.dart]: ditto

File: course/12_error/task_12_error/lib/unit_converter.dart [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+  /// Sets the default values for the 'from' and 'to' [Dropdown]s, and the
+  /// updated output value if a user had previously entered an input.
+  void _setDefaults() {
+    setState(() {
+      _fromValue = widget.category.units[0];
+      _toValue = widget.category.units[1];
+    });
+    if (_inputValue != null) {
+      setState(() {
```
Comment (yjbanov) [inline course/12_error/task_12_error/lib/unit_converter.dart]: ditto

**Comment by yjbanov** [comment]
lgtm

File: course/12_error/solution_12_error/lib/category.dart (hunk 1 above)
Comment (maryx) [reply course/12_error/solution_12_error/lib/category.dart]: Oops, forgot to add that one. Thanks.

File: course/12_error/solution_12_error/lib/category_route.dart (hunk 2 above)
Comment (maryx) [reply course/12_error/solution_12_error/lib/category_route.dart]: Thanks. Done.

File: course/12_error/solution_12_error/lib/unit_converter.dart (hunk 3 above)
Comment (maryx) [reply course/12_error/solution_12_error/lib/unit_converter.dart]: Done.

File: course/12_error/task_12_error/lib/unit_converter.dart (hunk 5 above)
Comment (maryx) [reply course/12_error/task_12_error/lib/unit_converter.dart]: Done.

File: course/12_error/task_12_error/lib/category.dart (hunk 4 above)
Comment (maryx) [reply course/12_error/task_12_error/lib/category.dart]: Done.

**Pull request merged**